cmd/experimental/migrate/internal/migrate.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2024 The Tessera authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/avast/retry-go/v4"
+	"github.com/transparency-dev/merkle/rfc6962"
+	"github.com/transparency-dev/trillian-tessera/api"
+	"github.com/transparency-dev/trillian-tessera/api/layout"
+	"github.com/transparency-dev/trillian-tessera/client"
+	"golang.org/x/sync/errgroup"
+	"k8s.io/klog/v2"
+)
+
+// copier controls the migration work.
+type copier struct {
+	// storage is the target we're migrating to.
+	storage    MigrationStorage
+	getEntries client.EntryBundleFetcherFunc
+
+	// sourceSize is the size of the source log.
+	sourceSize uint64
+	// sourceRoot is the root hash of the source log at sourceSize.
+	sourceRoot []byte
+
+	// todo contains work items to be completed.
+	todo chan bundle
+
+	// bundlesToMigrate is the total number of entry bundles which need to be copied.
+	bundlesToMigrate uint64
```

Comment:
Access to this field looks racy

Reply:
You're right - changed the approach

Reply:
It doesn't even need to be a field now. Can just be a local variable?

Reply:
done